Additional unbounded-array test that should continue to pass

tautschnig opened the pull request:
This test passes without nondet-propagation, and should continue to pass
if/when it is merged. Test provided by @romainbrenguier.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Additional unbounded-array test that should continue to pass

This test passes without nondet-propagation, and should continue to pass
if/when it is merged.

romainbrenguier commented on jbmc/regression/jbmc/nondet_propagation1/andNot_Pass.java:
```diff
@@ -0,0 +1,40 @@
+    b2.set(2);
+    b1.andNot(b2);
+
+    assert(!b1.get(0));
```
It would be good to have a similar test, but which shouldn't pass, so that we don't get a test passing because we under-approximated the behavior. The negation of this assertion could do. Also it wouldn't cost much to add assertions for the other bits: `assert(b1.get(1)); assert(!b1.get(2));`

tautschnig replied:
Done as suggested.

tautschnig commented:
Neither @romainbrenguier nor I can make this test fail locally, but it seems all CI set-ups agree that there's a problem. I had thought that only adding additional propagation (nondet propagation or field sensitivity) would surface the problem, but it seems it also does show up with current develop under circumstances not well understood.

Either way I'll try to fix the underlying problem, my current guess is an issue with byte updates.

tautschnig committed:
index_exprt flattening: ensure expression mapping is known

For unbounded arrays, convert_index creates free variables that should
subsequently be used when constructing equations in the array decision
procedure. This mapping is established via the conversion cache, which
will return an existing bit vector for a previously cached expression.
With byte-operator lowering, however, the expression cached did not
match the expression passed to the array decision procedure, and thus
breaking the mapping.

tautschnig commented:
@kroening Could you please take a look at the bugfix commit that I have added here? (I do nominally have all the approvals, but this commit is new and I'd like to have your opinion on whether I've gotten this right.)

tautschnig commented:
> Neither @romainbrenguier nor I can make this test fail locally, but it seems all CI set-ups agree that there's a problem.

JFTR: I had forgotten to add `BitSet.class` to the commit. Should now be fixed.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: ff2d3f36ab124fc3759ff0cf18413694a99d907c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105608489
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

martin-cs reviewed (approved):
The changes to `solvers` seem reasonable but I can't vouch for the Java tests.  This seems like something that should be trivially checkable in C and produce a large performance difference.

romainbrenguier commented on src/solvers/flattening/boolbv_index.cpp:
```diff
@@ -43,16 +43,21 @@ bvt boolbvt::convert_index(const index_exprt &expr)
       if(has_byte_operator(expr))
-        record_array_index(to_index_expr(lower_byte_operators(expr, ns)));
+      {
+        index_exprt final_expr = to_index_expr(lower_byte_operators(expr, ns));
```
nitpick: const

tautschnig replied:
Done.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9908dace4b8d411e632cff2392e2c1324b7a114f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105656963

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8bd75c8ccf94f5911d2c011d5b320a134e3c6857).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105796668
